File: java_gradle/src/main/java/the/primeagen/lexer/Token.java
package the.primeagen.lexer;

import java.util.Objects;

public class Token {

    private final Type m_type;
    private final String m_literal;

    public Token(Type type, String literal) {
        m_type = type;
        m_literal = literal;
    }

    public static Token of(Type type) {
        return new Token(type, null);
    }

    public static Token of(Type type, String literal) {
        return new Token(type, literal);
    }

    public Type getType() {
        return m_type;
    }

    public String getLiteral() {
        return m_literal;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Token token = (Token) o;
        return m_type == token.m_type && Objects.equals(m_literal, token.m_literal);
    }

    @Override
    public int hashCode() {
        return Objects.hash(m_type, m_literal);
    }

    @Override
    public String toString() {
        return "Token{" +
                "m_type=" + m_type +
                ", m_literal='" + m_literal + '\'' +
                '}';
    }

    public enum Type {
        IDENTIFIER,
        INT,
        EOF,
        ILLEGAL,
        EQUAL,
        PLUS,
        COMMA,
        SEMICOLON,
        LEFT_PARENTHESES,
        RIGHT_PARENTHESES,
        LEFT_BRACE,
        RIGHT_BRACE,
        FUNCTION,
        LET
    }
}

File: kotlin-deez/app/src/main/kotlin/jdsl/Token.kt
package jdsl

class Token(val type: TokenType, val literal: String? = null) {
    override fun toString(): String {
        return "$type $literal"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Token) return false

        if (type != other.type) return false
        return literal == other.literal
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + (literal?.hashCode() ?: 0)
        return result
    }
}